<a href="https://colab.research.google.com/github/JGPB0/CS/blob/main/Lab1/CS_Lab1_ParedesJimmy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cipher Fundamentals
## Paredes Jimmy

##Exercise 1: Base-64 encoding, hexadecimal representation, and modulus operator.

###1. Develop a web application that allows users to determine certain properties of numbers.
* Prime Number Checker:
  * Provide an input field where users can enter a number.
  * Implement a button that, when clicked, checks if the entered number is prime or not.
  * Display the result indicating whether the number is prime or not.
* GCD Calculator:
  * Present two input fields where users can enter two numbers.
  * Implement a button that, when clicked, calculates the Greatest Common Divisor (GCD) of the two entered numbers.
  * Display the GCD result to the user.

[Github repository with the code of the Web application](https://github.com/JGPB0/CS/blob/main/Lab1/WebApp.html)

###2. Determine the Base 64 , Hex and Binary values for the following strings:
  * Hola:
    * Base64: SG9sYQ==
    * Hexadecimal: 486f6c61
    * Binary: 0100100001101111011011000110000101100001
  * HOLA:
    * Base64: SE9MQQ==
    * Hexadecimal: 484f4c41
    * Binary: 01001000010011110100110001000001
  * HolA:
    * Base64: SG9sQQ==
    * Hexadecimal: 486f6c41
    * Binary: 01001000011011110110010001000001


###3. Determine the following ASCII strings for these encoded formats:
  * 45637561646F72
    * ASCII: Ecuador
  * 01011001 01000001 01000011 01001000 01000001 01011001 01010100 01000101 01000011 01001000
    * ASCII: YACHAYTECH
  * U2VndXJpZGFkIEluZm9ybT90aWNh
    * ASCII: Seguridad Inform?tica
  * RXNjdWVsYSBkZSBjaWVuY2lhcyBNYXRlbT90aWNhcyB5IENvbXB1dGFjaW9uYWxlcw==
    * ASCII: Escuela de ciencias Matem?ticas y Computacionales


###4. Using Python, what is the result of 7669 (mod 453)? Prove that this result is correct.

In [ ]:
print(f"The result of 7669(mod 453) is: {7669%453}")

The result of 7669(mod 453) is: 421


The result is indeed right because: 7669-16*453 = 421

## Exercise 2: Greatest common divisor GCD.

###1. Write a Python program to determine the GCD for the following:
  * 7001 and 10.
  * 4539 and 6.

In [ ]:
def find_gcd(a, b):
    while b:
        a, b = b, a % b
    return a

print(f"The GCD of 7001 and 10 is: {find_gcd(7001, 10)}")
print(f"The GCD of 4539 and 6 is: {find_gcd(4539, 6)}")

The GCD of 7001 and 10 is: 1
The GCD of 4539 and 6 is: 3


###2. Two numbers are co-prime if they do not share co-factors, apart from 1, which is gcd(a,b)=1. Determine if the following values are co-prime:
  * 5435 and 634
  * 5432 and 634

In [ ]:
def are_coprime(a, b):
    def find_gcd(x, y):
        while y:
            x, y = y, x % y
        return x

    gcd = find_gcd(a, b)
    return gcd == 1

def print_coprime(num1, num2):
    if are_coprime(num1, num2):
        print(f"{num1} and {num2} are co-prime.")
    else:
        print(f"{num1} and {num2} are not co-prime.")

num1 = 5435; num2 = 634; num3 = 5432

print_coprime(num1, num2)
print_coprime(num3, num2)

5435 and 634 are co-prime.
5432 and 634 are not co-prime.


##Exercise 3. Modulus and Exponentiation

###1. What is the result of the following:
 * 8¹³ mod(271)
 * 12²³ mod(973)

In [ ]:
print(f"8¹³ mod(271) is: {(8**13)%271}")
print(f"12²³ mod(973) is: {(12**23)%973}")

8¹³ mod(271) is: 119
12²³ mod(973) is: 514


###2. Implement a Python program which will determine the result of C = M<sup>e</sup> mod(p)
Prove the following:
  * message = 101, e=7, p = 293
  * message = 4, e=11, p = 79
  * message = 5, e=5, p = 53


In [ ]:
def expon_ciphers(m, e, p):
    c = (m ** e) % p
    return c

m1 = 101; e1 = 7;  p1 = 293
m2 = 4;   e2 = 11; p2 = 79
m3 = 5;   e3 = 5;  p3 = 53

print(f"The result of C with message=101, e=7 and p=293 is: {expon_ciphers(m1,e1,p1)}")
print(f"The result of C with message=4, e=11 and p=79 is: {expon_ciphers(m2,e2,p2)}")
print(f"The result of C with message=5, e=5 and p=53 is: {expon_ciphers(m3,e3,p3)}")

The result of C with message=101, e=7 and p=293 is: 176
The result of C with message=4, e=11 and p=79 is: 36
The result of C with message=5, e=5 and p=53 is: 51


###3. Implement the Python code given above and determine the highest prime number possible in the following ranges:
  * Up to 100
  * Up to 1000
  * Up to 5000
  * Up to 10000

In [1]:
from sympy import primerange

def find_highest_prime(limit):
    primes = list(primerange(2, limit + 1))
    return max(primes)

ranges = [100, 1000, 5000, 10000]

for limit in ranges:
    highest_prime = find_highest_prime(limit)
    print(f"Highest prime up to {limit}: {highest_prime}")

Highest prime up to 100: 97
Highest prime up to 1000: 997
Highest prime up to 5000: 4999
Highest prime up to 10000: 9973


###4. Which of the following numbers are prime numbers:
  * Is 858599509 prime? **Yes**
  * Is 982451653 prime? **Yes**
  * Is 982451652 prime? **No**

## Exercise 4. Random numbers

###1. Implement the Python code given above. Using: a=21, seed=35, c=31, and m=100, prove that the sequence gives 66 17 88 79 90.

In [ ]:
def rand_num(a, seed, c, m, num):
    x = []
    for _ in range(num):
        seed = (a*seed+c)%m
        x.append(seed)
    return x

a = 21; seed = 35; c = 31; m = 100; num = 5
print(f'The sequence with a={a}, seed={seed}, c={c} and m={m} is: {rand_num(a, seed, c, m, num)}')

The sequence with a=21, seed=35, c=31 and m=100 is: [66, 17, 88, 79, 90]


So, the code above proves that the sequence is obtained with the given data.


###2. Determine the sequence for: a=22, seed=35, c=31, and m=100. First four numbers of sequence?


In [ ]:
a = 22; seed = 35; c = 31; m = 100; num = 4
print(f'The first {num} numbers of the sequence with a={a}, seed={seed}, c={c} and m={m} is: {rand_num(a, seed, c, m, num)}')

The first 4 numbers of the sequence with a=22, seed=35, c=31 and m=100 is: [1, 53, 97, 65]


###3. Determine the sequence for: a=954365343, seed=436241, c=55119927, and m=1000000. First four numbers of sequence?

In [ ]:
a = 954365343; seed = 436241; c = 55119927; m = 1000000; num = 4
print(f'The first {num} numbers of the sequence with a={a}, seed={seed}, c={c} and m={m} is: {rand_num(a, seed, c, m, num)}')

The first 4 numbers of the sequence with a=954365343, seed=436241, c=55119927 and m=1000000 is: [715590, 917297, 157798, 514641]


###4. Determine the sequence for: a=2175143, seed=3553, c=10653, and m=1000000. First four numbers of sequence?

In [ ]:
a = 2175143; seed = 3553; c = 10653; m = 1000000; num = 4
print(f'The first {num} numbers of the sequence with a={a}, seed={seed}, c={c} and m={m} is: {rand_num(a, seed, c, m, num)}')

The first 4 numbers of the sequence with a=2175143, seed=3553, c=10653 and m=1000000 is: [293732, 114329, 934700, 172753]
